src/nvim/memory.c

```diff
@@ -576,41 +577,66 @@ void alloc_block(Arena *arena)
   blk->prev = prev_blk;
 }
 
+static int arena_align_offset(void *ptr, size_t alignment, size_t *offset)
+{
+  uintptr_t uptr = (uintptr_t)ptr;
+
+  if ((UINTPTR_MAX - uptr) < (alignment - 1)) {
+    return -1;
+  }
+
+  *offset = ((uptr + (alignment - 1)) & ~(alignment - 1)) - uptr;
+
+  return 0;
+}
+
 /// @param arena if NULL, do a global allocation. caller must then free the value!
 /// @param size if zero, will still return a non-null pointer, but not a unique one
 void *arena_alloc(Arena *arena, size_t size, bool align)
 {
   if (!arena) {
     return xmalloc(size);
   }
+  if (!arena->cur_blk) {
+    alloc_block(arena);
+  }
+  size_t align_pos = 0;
+  bool alignment_ok = true;
   if (align) {
-    arena->pos = (arena->pos + (ARENA_ALIGN - 1)) & ~(ARENA_ALIGN - 1);
+    if (arena_align_offset(arena->cur_blk + arena->pos, ARENA_ALIGN, &align_pos) < 0) {
+      alignment_ok = false;
+    }
   }
-  if (arena->pos + size > arena->size || !arena->cur_blk) {
-    if (size > (ARENA_BLOCK_SIZE - sizeof(struct consumed_blk)) >> 1) {
+  if (arena->pos + align_pos + size > arena->size || !alignment_ok) {
+    if (size + (align ? (ARENA_ALIGN - 1) : 0) > (ARENA_BLOCK_SIZE - sizeof(struct consumed_blk))
+        >> 1) {
```

Comment:
This looks needlessly complicated. Expressing the constraint and doing overflow checking direct in `arena->pos` integer space was much simpler (instead of first converting to pointer so we can play-pretend "secure pointer handling" by then converting back to uintptr to do pointer-space overflow check.). Remember, this is an interface for known internal consumers, not a malloc implementation which needs be c-standard paranoia compliant.

Reply:
To be clear you'd like to drop "[fix(memory): check for uintptr_t overflow](https://github.com/neovim/neovim/pull/20489/commits/df1bd1fcbf83e129ed8251433fd24459f94658c1)" correct?

Reply:
ye, that looks like it.